README.md
# Racing+ Rebalanced

This is the source code for [Racing+ Rebalanced](https://steamcommunity.com/sharedfiles/filedetails/?id=2263635610), a mod for the [Binding of Isaac: Afterbirth+](https://store.steampowered.com/app/401920/The_Binding_of_Isaac_Afterbirth/).

It is written in [TypeScript](https://www.typescriptlang.org/) using the [IsaacScript](https://isaacscript.github.io/) framework.


CHANGES.md
# Racing+ Rebalanced Mod Changes

<!-- markdownlint-disable MD033 -->

WARNING: Racing+ Rebalanced is a mod intended for speedrunners and people who play the game [competitively in actual tournaments](https://isaacrankings.com/). If you are a casual and/or don't play the game at a high level, then some of the balancing changes may not make sense to you. This mod is not recommended unless you are trying to complete the game as fast as possible (e.g. speedrunning or racing).

Racing+ Rebalanced is a separate mod that is intended to be run alongside the Racing+ mod. The vanilla game has many design flaws: it contains many completely useless items and it contains many overpowered items. Most important of all, it was never designed to be played competitively. The Rebalanced mod changes many things about the game in an attempt to:

1) Make every item in the game useful in a significant way / worth taking.
2) Slightly nerf the most powerful items.
3) Make the item pools more streamlined, each having a specific purpose.
4) Rework some mechanics in order to make the game more skill-based and reward players who take less hits.

<br>

## Table of Contents

1. [Characters](#characters)
1. [Items](#items)
1. [Trinkets](#trinkets)
1. [Cards](#cards)
1. [Pills](#pills)
1. [Room Revamps](#room-revamps)
1. [Miscellaneous](#miscellaneous)
1. [Starting Items](#starting-items)

<br>

## Characters

* All characters start with the Schoolbag.
* All characters start with Duality.
* Lilith now always starts with two Incubi (similar to the Season 4 custom challenge).
* The 7 characters used in the Season 9 custom challenge are:
  * Isaac
  * Magdalene
  * Cain
  * Judas
  * Blue Baby
  * Azazel
  * Lazarus

<br>

## Items

### Item Pool Revamp

* All items are now fully removed from their respective item pools on sight.
* All item pools are redesigned:
  * The Devil Room only contains offensive items.
  * The Angel Room only contains defensive items.
  * The shop contains only mapping and other utility items.
  * The Curse Room pool contains all of the 1 heart items from the Devil Room pool (with the exception of Cube of Meat).
* All items have weight 1.0 with the following exceptions:
  * The Compass, Treasure Map, Blue Map, and Book of Secrets are weighted higher such that there is a 15% chance to get one of them in the shop pool. (There is an average of 1.63 items per shop, which makes 24.5% prior to using a reroll machine.)
  * Cube of Meat is weighted higher such that there is a 15% chance to get it from the Devil Deal pool.
  * Ball of Bandages is weighted higher such that there is a 15% chance to get it from the Angel Room pool.
* The secret, golden chest, red chest, beggar, demon beggar, key master, and bomb bum pools are deleted. (These pools will now draw from the Treasure Room pool.)
* There are no longer set boss drops. (e.g. Steven, C.H.A.D, Gish, and all Horsemen will now drop a random item from the boss pool.)

### Item Removals

* Raw Liver (16) - This is not interesting enough to warrant being an item.
* A Dollar (18) - This item is too powerful with no good way to nerf it.
* Boom! (19) - This is not interesting enough to warrant being an item.
* Mom's Underwear (29) - Range ups are useless on the vast majority of speedruns.
* Mom's Heels (30) - Range ups are useless on the vast majority of speedruns.
* Mom's Lipstick (31) - Range ups are useless on the vast majority of speedruns.
* The Poop (36) - This is replaced with a new custom item called Holy Poop.
* Mom's Pad (41) - This item no longer does anything since fear is removed.
* Lucky Foot (46) - Luck ups are useless on the vast majority of speedruns.
* Mom's Eye (55) - This effect overlaps with Loki's Horns and buffing the latter is preferred.
* Lemon Mishap (56) - This effect overlaps with Bladder Infection.
* The Ladder (60) - With so many flight items in the game, it is not useful enough to warrant including in a pool.
* Charm of the Vampire (62) - Only beneficial on the Forgotten.
* The Hourglass (66) - Not useful enough to warrant including in a pool.
* Rosary (72) - The Bible / health trade off is strategically interesting, but not impactful enough to make the item useful enough to be a viable take.
* A Quarter (74) - While 25 coins are extremely useful in the improved shop pool, items should grant some unique effect other than just giving coins.
* Dead Cat (81) - This item is too powerful, allowing unskilled players to take 9 free devil deals.
* Lord of the Pit (82) - This effect overlaps with other items that grant flying.
* We Need To Go Deeper! (84) - This item is too powerful in speedruns, greatly decreasing the lower bound time in a carried run.
* Spider Bite (89) - This item is not interesting enough to warrant being an item.
* The Gamekid (93) - This effect overlaps with My Little Unicorn.
* Little C.H.A.D. (96) - Not useful enough to warrant including in a pool (since most builds are soul heart builds).
* The Common Cold (103) - This effect overlaps with Scorpio.
* The Bean (111) - This item is not interesting enough to warrant being an item.
* Guardian Angel (112) - The increased orbital speed makes this item only worth taking if you have no orbitals already. However, this trade-off no longer makes sense with the rebalanced orbital itemization.
* Odd Mushroom (Thin) (120) - This effect overlaps with Soy Milk.
* Dead Sea Scrolls (124) - A near-useless item with no clear way to buff it.
* Bobby-Bomb (125) - A frustrating item for speedruns.
* Forget Me Now (127) - Useless in speedruns.
* Bucket of Lard (129) - Near-useless in speedruns.
* Guppy's Paw (133) - This item removes the main skill-based mechanic of the game (e.g. red heart damage). And with Dead Cat removed, the main synergy is gone.
* Guppy's Tail (134) - With the Guppy transformation removed, the effect alone is not interesting enough to warrant being an item.
* Best Friend (136) - This effect overlaps with Mr. Boom!

src/pills.ts
import {
  CollectibleType,
  Direction,
  PillColor,
  PillEffect,
  RoomShape,
  SoundEffect,
} from "isaac-typescript-definitions";
import { sfxManager } from "isaacscript-common";
import { PillEffectCustom } from "./enums/PillEffectCustom";
import g from "./globals";

const DURATION = 600; // 20 seconds

export const COLORS = [
  PillColor.BLUE_BLUE, // 1
  PillColor.ORANGE_ORANGE, // 3
  PillColor.WHITE_WHITE, // 4
  PillColor.RED_DOTS_RED, // 5 (the sprite for this is changed to a full red pill in Racing+)
];

export const EFFECTS = [
  // Vanilla pills
  PillEffect.BALLS_OF_STEEL, // 2
  PillEffect.HEALTH_UP, // 7
  PillEffect.PRETTY_FLY, // 10
  PillEffect.SPEED_UP, // 14
  PillEffect.TEARS_UP, // 16
  PillEffect.FORTY_EIGHT_HOUR_ENERGY, // 20
  PillEffect.I_CAN_SEE_FOREVER, // 23

  // Custom pills
  PillEffectCustom.DAMAGE_UP,
  PillEffectCustom.TEAR_DELAY_DOWN,
  PillEffectCustom.DEAL_AFFINITY,
  PillEffectCustom.BONE_AFFINITY,
  PillEffectCustom.RESTOCK,
  PillEffectCustom.GOLDEN_DUMP,
  PillEffectCustom.SUPER_SADNESS,
  PillEffectCustom.INVINCIBILITY,
  PillEffectCustom.REALLY_BAD_GAS,
  PillEffectCustom.GLIMPSE,
  PillEffectCustom.AETHER,
  PillEffectCustom.WALLS_HAVE_EYES,
  PillEffectCustom.BLADDER_INFECTION,
  PillEffectCustom.SCORCHED_EARTH,
];

// - The first number is the starting grid coordinate.
// - The second number is the number of times to iterate.
// - The third number is the starting grid coordinate for the second wall (for L rooms).
export const ROOM_SHAPE_WALL_COORDINATES: {
  readonly [key in RoomShape]: Map<
    Direction,
    [
      startingGridCoordinate: int,
      numTimesToIterate: int,
      startingGridCoordinateForSecondWall?: int,
    ]
  >;
} = {
  // 1
  [RoomShape.SHAPE_1x1]: new Map([
    [Direction.LEFT, [29, 7]],
    [Direction.UP, [121, 13]],
    [Direction.RIGHT, [15, 7]],
    [Direction.DOWN, [1, 13]],
  ]),

  // 2
  [RoomShape.IH]: new Map([
    [Direction.LEFT, [59, 3]],
    [Direction.UP, [91, 13]],
    [Direction.RIGHT, [45, 3]],
    [Direction.DOWN, [31, 13]],
  ]),

  // 3
  [RoomShape.IV]: new Map([
    [Direction.LEFT, [25, 7]],
    [Direction.UP, [125, 5]],
    [Direction.RIGHT, [19, 7]],
    [Direction.DOWN, [5, 5]],
  ]),

  // 4
  [RoomShape.SHAPE_1x2]: new Map([
    [Direction.LEFT, [44, 14]],
    [Direction.UP, [226, 13]],
    [Direction.RIGHT, [15, 14]],
    [Direction.DOWN, [1, 13]],
  ]),

  // 5
  [RoomShape.IIV]: new Map([
    [Direction.LEFT, [25, 14]],
    [Direction.UP, [230, 5]],
    [Direction.RIGHT, [19, 14]],
    [Direction.DOWN, [5, 5]],
  ]),

  // 6
  [RoomShape.SHAPE_2x1]: new Map([
    [Direction.LEFT, [55, 7]],
    [Direction.UP, [225, 26]],
    [Direction.RIGHT, [28, 7]],
    [Direction.DOWN, [1, 26]],
  ]),

  // 7
  [RoomShape.IIH]: new Map([
    [Direction.LEFT, [111, 3]],
    [Direction.UP, [169, 26]],
    [Direction.RIGHT, [84, 3]],
    [Direction.DOWN, [57, 26]],
  ]),

  // 8
  [RoomShape.SHAPE_2x2]: new Map([
    [Direction.LEFT, [55, 14]],
    [Direction.UP, [421, 26]],
    [Direction.RIGHT, [28, 14]],
    [Direction.DOWN, [1, 26]],
  ]),

  // 9
  [RoomShape.LTL]: new Map([
    [Direction.LEFT, [55, 14]],
    [Direction.UP, [421, 26]],
    [Direction.RIGHT, [41, 7, 224]],
    [Direction.DOWN, [197, 13, 14]],
  ]),

  // 10
  [RoomShape.LTR]: new Map([
    [Direction.LEFT, [42, 7, 251]],
    [Direction.UP, [421, 26]],
    [Direction.RIGHT, [28, 14]],
    [Direction.DOWN, [1, 13, 210]],
  ]),

  // 11
  [RoomShape.LBL]: new Map([
    [Direction.LEFT, [55, 14]],
    [Direction.UP, [225, 13, 434]],
    [Direction.RIGHT, [28, 7, 237]],
    [Direction.DOWN, [1, 26]],
  ]),

  // 12
  [RoomShape.LBR]: new Map([
    [Direction.LEFT, [55, 7, 238]],
    [Direction.UP, [421, 13, 238]],
    [Direction.RIGHT, [28, 14]],
    [Direction.DOWN, [1, 26]],
  ]),
} as const;

export function animateHappy(): void {
  g.p.AnimateHappy();
  const color = Color(0.3, 0.3, 0.3, 1, 1, 1, 1);
  g.p.SetColor(color, 15, 1, true, false);
  sfxManager.Stop(SoundEffect.THUMBS_UP);
  sfxManager.Play(SoundEffect.POWER_UP_SPEWER, 1, 0, false, 1);
}

export function getDuration(): int {
  let duration = DURATION;
  if (g.p.HasCollectible(CollectibleType.PHD)) {
    duration *= 2;
  }

  return duration;
}
